Make Downward API test table-driven

ivan4th opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35055)

<!-- Reviewable:end -->

Commit:
Make Downward API test table-driven

ivan4th commented:
@k8s-bot gci gke e2e test this

ivan4th commented:
@k8s-bot gci gke e2e test this

ivan4th commented:
(ping)

ivan4th commented:
(ping)

pmorie commented:
@ivan4th sorry for the latency, reviewing now.

pmorie commented:
@ivan4th The only thing I would change about this is move the verify functions down to be next to the testStep type.  Do that and I will merge.  Good patch.

ivan4th commented:
@pmorie fixed. I kept `testStep`'s `run` method close to it though.

ivan4th commented:
@k8s-bot gce etcd3 e2e test this

ivan4th commented:
@k8s-bot gce e2e test this

ivan4th commented:
@k8s-bot gce node e2e test this

ivan4th commented:
@k8s-bot node e2e test this

ivan4th commented:
also ping...

ivan4th commented:
Did the requested changes.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35055#-:-KYo3BjWBH20f7JxPTdQ:b9b1asy)*
<!-- Sent from Reviewable.io -->

ivan4th commented:
(oops, commented the wrong PR, this one has no requested changes)

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35055#-:-KYo3cXqckePgoMpQX7f:buqssy1)*
<!-- Sent from Reviewable.io -->

ivan4th commented:
(ping)

ivan4th commented:
(ping)

saad-ali commented:
Sorry for the delay @ivan4th. And thanks for the refactor!

saad-ali commented:
@rkouj could you take a look please

rkouj commented:
Ack

rkouj commented:
This is an XL PR and is taking more time than expected. I'm working on it.

rkouj commented on pkg/volume/downwardapi/downwardapi_test.go:
```diff
@@ -493,346 +228,131 @@ func TestWriteTwiceWithUpdate(t *testing.T) {
-	if err != nil {
-		t.Errorf(err.Error())
-	}
+type testStep struct {
```
Please add a comment on top of this as it is unclear what this is supposed to be.

rkouj commented on pkg/volume/downwardapi/downwardapi_test.go:
```diff
@@ -60,431 +57,169 @@ func TestCanSupport(t *testing.T) {
+			files:     map[string]string{"labels": "metadata.labels"},
+			podLabels: labels1,
+			steps: []testStep{
+				{name: "<resetup>"},
```
Is there a special meaning to this string with this syntax?

rkouj commented on pkg/volume/downwardapi/downwardapi_test.go:
```diff
@@ -493,346 +228,131 @@ func TestWriteTwiceWithUpdate(t *testing.T) {
-	if sortLines(string(data)) != sortLines(fieldpath.FormatMap(labels)) {
-		t.Errorf("Found `%s` expected %s", data, fieldpath.FormatMap(labels))
+func (s testStep) run(test *downwardAPITest) {
+	switch {
```
Please add comments for each of the cases

rkouj replied:
This block is fragile and is heavily dependent on the order.

Someone who is making changes to the existing tests if accidentally sets/unsets one of the variables above, then we can get unexpected behavior because of the ordering and dependence of a verification function on other unrelated factors.

I'd suggest working on a different approach for handling this. Either you can have a map with the testName as key and the verification function as value or something else on similar lines.

rkouj reviewed (changes requested):
Thanks for the refactor!

Please don't change the order of functions if you plan on keeping them. For example the function newTestHost() is deleted and re-created at the bottom. Makes it hard for the reviewer to determine what actually changed.

Please consider the following comments.

ivan4th commented:
`rkouj was unassigned by ivan4th 20 seconds ago` -- hmm I didn't do this

Also, sorry for the delay.

As of the changes: I replaced switch-based approach with less fragile one with different struct types serving as test steps. I also moved `newTestHost()` back on top. The initial motivation for moving it to bottom was that many `*_test.go` files in k8s are organized this way, with test themselves on top and helper function on the bottom.

saad-ali commented:
Thanks for the cleanup @ivan4th

saad-ali commented:
/lgtm
/approve

ivan4th commented:
@k8s-bot test this

ivan4th commented:
@k8s-bot test this
